### ISSUE TITLE
`noarch: python` entry points do not work on Windows if the prefix path contains certain Unicode symbols

### ISSUE THREAD

**Issue by jaimergp** [conda/conda#11694]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

conda has `noarch: python` packages to help reuse the same artifact across Python versions and operating systems. This requires some clever workarounds like regenerating the entry points on Unix when the Python version changes (e.g. 3.6 -> 3.7) so the shebang in the entry point gets updated to the new Python executable.

On Windows, entry points work a bit differently. Since there's no concept of shebang, Windows reuses the Python launcher used by `conda` itself. This launcher is a simple binary that locates a Python executable and passes a `*-script.py` file located next to the launcher.

I _think_ this is provided simply by copying the `conda.exe` launcher (in `CONDA_ROOT/condabin/`) to `PREFIX/Scripts` with the entry point name. For example, an entry point like `flask=flask.cli:main` will generate two files:

- `flask.exe`: the copy of `condabin/conda.exe`
- `flask-script.py` the Unix equivalent entry point.

So when the user runs `flask` in their activated conda environment, `flask.exe` locates the environment `python.exe` and passes the `flask-script.py` file.

> Where's the original `conda.exe` coming from? 

Where's the problem? There must be a bug either in the launcher we are using, because if the environment path contains Unicode symbols, the script is not correctly found or passed:

```
$ conda create -p "./environment 0b74 한áō固ђγ家êôñßç" conda-forge::flask
# flask on defaults is not noarch, conda-forge's is
$ cd "./environment 0b74 한áō固ђγ家êôñßç"
$ conda activate "./"
$ flask
Cannot open D:\Users\jrodriguez\a_local_channel 0b74 ?�o????�����\Scripts\flask-script.py
```

What I see is a Unicode encode or decode issue.

### Conda Details

<details>
<summary><code>conda info</code></summary>

```shell
active environment : base
    active env location : D:\Users\jrodriguez\Miniconda3
            shell level : 1
       user config file : D:\Users\jrodriguez\.condarc
 populated config files :
          conda version : 4.13.0
    conda-build version : 3.21.8
         python version : 3.9.5.final.0
       virtual packages : __win=0=0
                          __archspec=1=x86_64
       base environment : D:\Users\jrodriguez\Miniconda3  (writable)
      conda av data dir : D:\Users\jrodriguez\Miniconda3\etc\conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/win-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/win-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/msys2/win-64
                          https://repo.anaconda.com/pkgs/msys2/noarch
          package cache : D:\Users\jrodriguez\Miniconda3\pkgs
                          D:\Users\jrodriguez\.conda\pkgs
                          D:\Users\jrodriguez\AppData\Local\conda\conda\pkgs
       envs directories : D:\Users\jrodriguez\Miniconda3\envs
                          D:\Users\jrodriguez\.conda\envs
                          D:\Users\jrodriguez\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/4.13.0 requests/2.25.1 CPython/3.9.5 Windows/10 Windows/10.0.14393
          administrator : False
             netrc file : None
           offline mode : False
```

</details>
<details>
<summary><code>conda list</code></summary>

### Conda list

```shell
# packages in environment at D:\Users\jrodriguez\Miniconda3:
#
# Name                    Version                   Build  Channel
anaconda-client           1.8.0              pyhd8ed1ab_0    conda-forge
anyio                     3.5.0            py39hcbf5309_0    conda-forge
atomicwrites              1.4.0              pyh9f0ad1d_0    conda-forge
attrs                     21.4.0             pyhd8ed1ab_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
beautifulsoup4            4.11.1             pyha770c72_0    conda-forge
boa                       0.10.0             pyha770c72_0    conda-forge
brotlipy                  0.7.0           py39h2bbff1b_1003    defaults
bs4                       4.11.1               hd8ed1ab_0    conda-forge
bzip2                     1.0.8                h8ffe710_4    conda-forge
ca-certificates           2022.07.19           haa95532_0    defaults
certifi                   2022.6.15        py39haa95532_0    defaults
cffi                      1.14.6           py39h2bbff1b_0    defaults
chardet                   4.0.0            py39hcbf5309_3    conda-forge
clyent                    1.2.2                      py_1    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
commonmark                0.9.1                      py_0    conda-forge
conda                     4.13.0           py39haa95532_0    defaults
conda-build               3.21.8           py39hcbf5309_0    conda-forge
conda-libmamba-solver     22.3.1             pyhd8ed1ab_0    conda-forge
conda-package-handling    1.8.1            py39hb3671d1_1    conda-forge
console_shortcut          0.1.1                         4    defaults
coverage                  6.3.2            py39hb82d6ee_2    conda-forge
cryptography              3.4.7            py39h71e12ea_0    defaults
dataclasses               0.8                pyhc8e2a94_3    conda-forge
docutils                  0.18.1           py39hcbf5309_1    conda-forge
filelock                  3.6.0              pyhd8ed1ab_0    conda-forge
flaky                     3.7.0              pyh9f0ad1d_0    conda-forge
flit                      3.7.1              pyhd8ed1ab_0    conda-forge
flit-core                 3.7.1              pyhd8ed1ab_0    conda-forge
future                    0.18.2           py39hcbf5309_5    conda-forge
git                       2.34.1               haa95532_0    defaults
glob2                     0.7                        py_0    conda-forge
idna                      2.10               pyhd3eb1b0_0    defaults
importlib-metadata        4.11.3           py39hcbf5309_1    conda-forge
importlib_resources       5.7.1              pyhd8ed1ab_0    conda-forge
iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge
jinja2                    3.1.1              pyhd8ed1ab_0    conda-forge
joblib                    1.1.0              pyhd8ed1ab_0    conda-forge
json5                     0.9.5              pyh9f0ad1d_0    conda-forge
jsonschema                4.4.0              pyhd8ed1ab_0    conda-forge
jupyter_core              4.9.2            py39hcbf5309_0    conda-forge
krb5                      1.19.2               h20d022d_3    conda-forge
libarchive                3.5.2                hb45042f_1    conda-forge
libcurl                   7.81.0               h789b8ee_0    conda-forge
libiconv                  1.16                 he774522_0    conda-forge
liblief                   0.11.5               h0e60522_1    conda-forge
libmamba                  0.22.1               h44daa3b_0    conda-forge
libmambapy                0.22.1           py39h9c8ecf5_0    conda-forge
libsolv                   0.7.19               h7755175_5    conda-forge
libssh2                   1.10.0               h680486a_2    conda-forge
libxml2                   2.9.12               hf5bbc77_1    conda-forge
libzlib                   1.2.11            h8ffe710_1013    conda-forge
lz4-c                     1.9.3                h8ffe710_1    conda-forge
lzo                       2.10              he774522_1000    conda-forge
m2-msys2-runtime          2.5.0.17080.65c939c               3    conda-forge
m2-patch                  2.7.5                         2    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
mamba                     0.22.1           py39hb3d9227_0    conda-forge
markupsafe                2.1.1            py39hb82d6ee_1    conda-forge
menuinst                  1.4.16           py39h2bbff1b_0    defaults
msys2-conda-epoch         20160418                      1    conda-forge
nbformat                  5.3.0              pyhd8ed1ab_0    conda-forge
openssl                   1.1.1q               h2bbff1b_0    defaults
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pip                       21.1.3           py39haa95532_0    defaults
pkginfo                   1.8.2              pyhd8ed1ab_0    conda-forge
pluggy                    1.0.0            py39hcbf5309_3    conda-forge
powershell_shortcut       0.0.1                         3    defaults
prompt-toolkit            3.0.29             pyha770c72_0    conda-forge
prompt_toolkit            3.0.29               hd8ed1ab_0    conda-forge
psutil                    5.9.0            py39hb82d6ee_1    conda-forge
py                        1.11.0             pyh6c4a22f_0    conda-forge
py-lief                   0.11.5           py39h415ef7b_1    conda-forge
pybind11-abi              4                    hd8ed1ab_3    conda-forge
pycosat                   0.6.3            py39h2bbff1b_0    defaults
pycparser                 2.20                       py_2    defaults
pygments                  2.11.2             pyhd8ed1ab_0    conda-forge
pyopenssl                 20.0.1             pyhd3eb1b0_1    defaults
pyparsing                 3.0.8              pyhd8ed1ab_0    conda-forge
pyrsistent                0.18.1           py39hb82d6ee_1    conda-forge
pysocks                   1.7.1            py39haa95532_0    defaults
pytest                    7.1.1            py39hcbf5309_1    conda-forge
pytest-cov                3.0.0              pyhd8ed1ab_0    conda-forge
pytest-mock               3.7.0              pyhd8ed1ab_0    conda-forge
python                    3.9.5                h6244533_3    defaults
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-fastjsonschema     2.15.3             pyhd8ed1ab_0    conda-forge
python-libarchive-c       4.0              py39hcbf5309_1    conda-forge
python_abi                3.9                      2_cp39    conda-forge
pytz                      2022.1             pyhd8ed1ab_0    conda-forge
pywin32                   228              py39he774522_0    defaults
pyyaml                    6.0              py39hb82d6ee_4    conda-forge
reproc                    14.2.3               h8ffe710_0    conda-forge
reproc-cpp                14.2.3               h0e60522_0    conda-forge
requests                  2.25.1             pyhd3eb1b0_0    defaults
rich                      12.2.0             pyhd8ed1ab_0    conda-forge
ripgrep                   13.0.0               h7f3b576_2    conda-forge
ruamel.yaml               0.17.21          py39hb82d6ee_1    conda-forge
ruamel.yaml.clib          0.2.6            py39hb82d6ee_1    conda-forge
ruamel_yaml               0.15.100         py39h2bbff1b_0    defaults
setuptools                52.0.0           py39haa95532_0    defaults
six                       1.16.0             pyhd3eb1b0_0    defaults
sniffio                   1.2.0            py39hcbf5309_3    conda-forge
soupsieve                 2.3.1              pyhd8ed1ab_0    conda-forge
sqlite                    3.36.0               h2bbff1b_0    defaults
tk                        8.6.12               h8ffe710_0    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
tomli-w                   1.0.0              pyhd8ed1ab_0    conda-forge
tqdm                      4.61.2             pyhd3eb1b0_1    defaults
traitlets                 5.1.1              pyhd8ed1ab_0    conda-forge
typing_extensions         4.2.0              pyha770c72_1    conda-forge
tzdata                    2021a                h52ac0ba_0    defaults
urllib3                   1.26.6             pyhd3eb1b0_1    defaults
vc                        14.2                 h21ff451_1    defaults
vs2015_runtime            14.27.29016          h5e58377_2    defaults
watchgod                  0.8.2              pyhd8ed1ab_0    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
wheel                     0.36.2             pyhd3eb1b0_0    defaults
win_inet_pton             1.1.0            py39haa95532_0    defaults
wincertstore              0.2              py39h2bbff1b_0    defaults
xz                        5.2.5                h62dcd97_1    conda-forge
yaml                      0.2.5                he774522_0    defaults
yaml-cpp                  0.6.3                ha925a31_4    conda-forge
zipp                      3.8.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11            h8ffe710_1013    conda-forge
zstd                      1.5.2                h6255e5f_0    conda-forge
```

</details>

### Additional Context

Ok, so there's an error with the launcher. Question is... where does the `conda.exe` launcher come from?

Option A:
* The launchers for each `conda` release are bundled in `conda-build` itself (`{cli,gui}-{32,64}.exe`), and they are generated with [these sources](https://github.com/conda/conda-build/tree/main/conda_build/launcher_sources).
*  Last time they were updated: 2020 via https://github.com/conda/conda-build/pull/3894
* That means maybe the bug is more on the `conda-build` side? No clue yet.

Option B: 
* `conda/conda` also has [their own `cli-{32,64}.exe` files in the repo](https://github.com/conda/conda/tree/main/conda/shell), last updated 4 years ago. Are these used?
* Note that the `setuptools`-generated installers are [deliberately deleted](https://github.com/conda-forge/conda-feedstock/blob/78cc35ef3614df48de1fab2e3eeb83b0c85fd565/recipe/build.sh#L15) in the feedstock. Instead, it later relies on `python -m conda init --install`.
* This code path is in `conda.core.initialize`! There, the _plan_ uses [this function](https://github.com/conda/conda/blob/20ea9d5a7a658ab95acd2abcd5b7484f11686110/conda/core/initialize.py#L789) to generate its own entry points. 
* A-ha! It uses `source_exe_path = join(CONDA_PACKAGE_ROOT, 'shell', 'cli-%d.exe' % bits)` -- aka the vendored `cli-*.exe` files we can find in this repo! So... do we need to update them?

So what are the EXEs in `conda-build` for? Looks like conda-build [duplicates a lot of this logic too](https://github.com/conda/conda-build/blob/42b84ee27ff0f51e50b449b424f69ca8a7ffe8e0/conda_build/_link.py#L94), so maybe they need it in case conda is not around, or maybe it's technical debt?

> Note: If `flask` is installed with `pip`, it does work. This means that the [`setuptools` launcher](https://github.com/pypa/setuptools/blob/main/launcher.c) works! Their approach is essentially the same.

**Comment by jaimergp** [comment]
Went down the rabbit hole and learnt a bit more about this.

`conda-build` vendors its own launchers because it can create arbitrary entry points for non-noarch packages too. It does so at build time. `conda` needs them because it can create `noarch` entry points at install time.

I realized two things:

* `conda-build`'s launchers are always 41KB on Windows 64 bit, but `conda`'s are always 73KB. The most recent ones are the ones in conda-build (3 years old in CB vs 4 years old in conda).
* Since they are all copies of the same files, you can move them around and rename them accordingly. The 41KB versions WORKS on Unicode envs. The 73KB doesn't.

In conclusion, all we need to do is copy over the `conda-build` cli launchers :D  This will also allow `conda` to be installed in Unicode-containing environments on Windows (if they relied on the `conda.exe` entry point).
